plugins/ssh-plugin/src/git/git-listener.ts

```diff
@@ -8,107 +8,105 @@
+import { SSHPlugin } from '../plugin/plugin-model';
+
+const RETRY = 'Retry';
+const ADD_KEY_TO_GITHUB = 'Add Key To GitHub';
+const CONFIGURE_SSH = 'Configure SSH';
 
 @injectable()
 export class GitListener {
-  @inject(AddKeyToGitHub)
-  private addKeyToGitHub: AddKeyToGitHub;
+  @inject(SSHPlugin)
+  private sshPlugin: SSHPlugin;
 
-  @inject(GenerateKey)
-  private generateKey: GenerateKey;
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  private git: any;
 
-  @inject(ViewPublicKey)
-  private viewPublicKey: ViewPublicKey;
+  private gitCommand: string = '';
 
   constructor() {}
 
-  private async getKeys(): Promise<cheApi.ssh.SshPair[]> {
-    let keys: cheApi.ssh.SshPair[];
-    try {
-      keys = await che.ssh.getAll('vcs');
-    } catch (e) {
-      console.error(e && e.message ? e.message : e);
-      keys = [];
+  private async onGitOutput(data: string): Promise<void> {
+    if (data.startsWith('> git clone') || data.startsWith('> git push')) {
+      this.gitCommand = data;
+      return;
+    }
+
+    if (data.indexOf('Permission denied (publickey).') < 0) {
+      // In this listener we handle only issues with SSH key
+      return;
     }
 
-    return keys;
+    // Parse Git log events.
+    if (this.gitCommand.startsWith('> git clone') || this.gitCommand.startsWith('> git push')) {
+      const split = this.gitCommand.split(' ');
+      const command = split[2];
+      const uri = split[3];
+      const path = split[4];
+
+      if (command === 'clone') {
+        await this.handleGitClone(uri, path);
+      } else if (command === 'push') {
+        await this.handleGitPush(uri, path);
+      }
+    }
   }
 
-  async init() {
-    const keys = await this.getKeys();
+  private isGitHubUri(uri: string): boolean {
+    return uri.startsWith('git@github.com');
+  }
+
+  private async handleGitClone(uri: string, path: string): Promise<void> {
+    const message = `Failure to clone git project ${uri}. A valid SSH key may be required.`;
+
+    const gitHub = this.isGitHubUri(uri);
+    const buttons = gitHub ? [RETRY, ADD_KEY_TO_GITHUB, CONFIGURE_SSH] : [RETRY, CONFIGURE_SSH];
 
-    let gitLogHandlerInitialized: boolean;
-    /* Git log handler, listens to Git events, catches the clone and push events.
-          Asks to Upload a public SSH key if needed before these operations.
-          Authenticates to Github if needed. */
+    const action = await theia.window.showWarningMessage(message, ...buttons);
+    if (action === RETRY) {
+      await this.retryClone(uri, path);
+    } else if (action === ADD_KEY_TO_GITHUB) {
+      await this.sshPlugin.addKeyToGitHub();
+      await this.retryClone(uri, path);
+    } else if (action === CONFIGURE_SSH) {
+      await this.sshPlugin.configureSSH(gitHub);
+      await this.retryClone(uri, path);
+    }
+  }
+
+  private async retryClone(uri: string, path: string) {
+    await this.git.clone(uri, path.substring(0, path.lastIndexOf('/')));
```

Comment:
Please add a progress notification so that user can see what happens. The same logic is present in the workspace plugin: https://github.com/eclipse-che/che-theia/blob/c93df2ac846a95fe7252caa191779773c372035a/plugins/workspace-plugin/src/theia-commands.ts#L94-L108

Reply:
It's redundant here as we only handle an errors related to SSH and call the gist extension to re-clone the project.

Reply:
When I try to clone via ssh url and then upload ssh key with the help of notification the clone process is hidden.

Reply:
Done